massgov/mds-licensing
# Refactor ETLs to work per-environment

This PR reimplements a large chunk of the ETL process.  ETLs used to flow like this:

* [Cron from drop server] CSV file is uploaded to S3
* [csv_parser Lambda triggered by S3 upload] CSV file is converted to JSON file in S3
* [load_external_db_mf Lambda triggered by cron] JSON is read from S3 and pushed into Postgres `stage` database in agency-specific tables.
* [mf_PostGRES_Function Lambda triggered by cron] Data is combined in `stage` database - take agency specific tables and merge them into a single licensing table.
* [mf_PostGRES_Function Lambda triggered by cron] Data is copied from `stage.licensing` to `uat.licensing`
* [mf_PostGRES_Function Lambda triggered by cron] Data is copied from `stage.licensing` to `public.licensing`

After this PR, they flow like this:

* [Cron from drop server] CSV file is uploaded to S3
* [csv_parser Lambda triggered by S3 upload] CSV file is converted to JSON file in S3
* [licensing-dev-etl Lambda triggered by cron with `dev` as argument] For the `dev` schema, JSON is read from S3, pushed into a single `licensing_stage` table, verified, then the `licensing-stage` data will be copied over to `licensing`, replacing existing data.
* [licensing-test-etl Lambda triggered by cron with `test` as argument] For the `test` schema, JSON is read from S3, pushed into a single `licensing_stage` table, verified, then the `licensing-stage` data will be copied over to `licensing`, replacing existing data.
* [licensing-prod-etl Lambda triggered by cron with `prod` as argument] For the `prod` schema, JSON is read from S3, pushed into a single `licensing_stage` table, verified, then the `licensing-stage` data will be copied over to `licensing`, replacing existing data. 

(and in fact Mark pointed out that we may not want to set up the dev and test Lambdas with cron at all, so that might change).  

This Lambda was designed to be fault-tolerant.  If the day's data is not dropped off before the ETL runs, it will simply grab yesterday's data and fail the Lambda after the data has been imported.  If the process fails mid-stream for any reason, yesterday's data will not be overwritten.  If any agency has < 1000 records, yesterday's data will not be overwritten.  Once this is wired into alerts, it should give us a very clear picture of when things are not working.


Testing Steps:
* Run the `licensing-dev-v1-etl` lambda and ensure it populates the licenses table in the `dev` schema of the `licensing` database.
* Try to break processing in all sorts of weird ways.
